A MemoryManager for the PageCache

chrisvest opened the pull request:
The idea is to be able to allocate the cache pages at any alignment that the PageSwappers might require.
Useful if we want to start doing direct-io.

chrisvest committed:
Add first stab at a MemoryManager for Muninn

The idea is to be able to allocate the cache pages at any alignment that the PageSwappers might require.
Useful if we want to start doing direct-io.

chrisvest committed:
Complete and test the UnsafeUtil primitive accessor API

thobe commented on community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MemoryManager.java:
```diff
@@ -0,0 +1,134 @@
+    {
+        if ( slabs == null || !slabs.canAllocate( bytes ) )
+        {
+            long slabGrab = Math.min( GRAB_SIZE, memoryReserve );
```
There are no checks for `memoryReserve` being low, instead what will happen here is that this `min` will return `memoryReserve` once it dips below `GRAB_SIZE`, it might even return something negative, meaning that we will get a very strange `slabGrab`, one that might be significantly smaller than `bytes`, at which point just adding `alignment` isn't going to suffice.

chrisvest replied:
Nice catch.

chrisvest replied:
No, hang on. We are adding the alignment to _bytes_ and this sum overrides whatever slabGrab was before. The memoryReserver will go negative, and that's fine, because we don't use it in the calculations if it's less than bytes.

thobe replied:
Ah, right, it gets added to bytes... Then that part is strange... Why does alignment need to be added there? What is to say that bytes isn't aligned already?

chrisvest replied:
The algorithm aligns the pointers by allocating excess space, so the pointer returned by the memory manager can be rounded up to the next alignment point.

thobe replied:
I'd think there are cases, perhaps several cases, where this would lead to quite a bit of excess being allocated, even though everything would have been perfectly aligned.

chrisvest replied:
This is only relevant when the memoryReserve has been exhausted. And then only on configurations where large alignments are required. Would you rather we first try to allocate the exact number of requested bytes, and only fall back to this strategy if the pointer we get back isn't aligned?

thobe replied:
What I would like is for us to know what the alignment of the underlying `malloc` is, and utilise that knowledge, but if we can't do that the current implementation is ok-ish since it doesn't require recomputation in the common case.

chrisvest replied:
I don't think we can reliably probe the alignment of malloc. The specification says "The pointer returned if the allocation succeeds shall be suitably aligned so that it may be assigned to a pointer to any type of object", but then we're not guaranteed that Unsafe.allocateMemory is synonymous with malloc.

chrisvest committed:
Don't allocate more memory than necessary, if normal allocation is already aligned

chrisvest committed:
Rename UnsafeUtil.malloc to allocateMemory

This better reflects reality, because `sun.misc.Unsafe.allocateMemory` is not guaranteed to be synonymous with `malloc(3)`.

chrisvest commented:
@thobe @digitalstain The MemoryManager now no longer allocates more memory than strictly necessary, if the "native" alignment is already good enough.

thobe commented:
Oh, so now it optimistically assumes that the native allocator is going to provide good allocation, and then reallocates if that turns out to be false. That works, but could of course turn out to be an extra allocate+free if that assumption doesn't hold.

I think you should add a tracer for the allocations so that we can get statistics for how often we have to perform such reallocation in reality. We might be able to use that for inspecting how much space we waste by not being able to utilise full slabs and things like that.
